substrate/frame/migrations/src/tests.rs

```diff
@@ -0,0 +1,369 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg(test)]
+
+use crate::{
+	mock::{MockedMigrationKind::*, Test as T, *},
+	Cursor, Event, FailedUpgradeHandling, MigrationCursor,
+};
+use frame_support::{pallet_prelude::Weight, traits::OnRuntimeUpgrade};
+
+#[test]
+#[docify::export]
+fn simple_works() {
+	use Event::*;
+	test_closure(|| {
+		// Add three migrations, each taking one block longer than the last.
+		MigrationsStorage::set(vec![(SucceedAfter, 0), (SucceedAfter, 1), (SucceedAfter, 2)]);
+
+		System::set_block_number(1);
+		Migrations::on_runtime_upgrade();
+		run_to_block(10);
+
+		// Check that the executed migrations are recorded in `Historical`.
+		assert_eq!(
+			historic(),
+			vec![
+				mocked_id(SucceedAfter, 0),
+				mocked_id(SucceedAfter, 1),
+				mocked_id(SucceedAfter, 2),
+			]
+		);
+		// Check that we got all events.
+		assert_events(vec![
+			UpgradeStarted { migrations: 3 },
+			MigrationCompleted { index: 0, blocks: 1 },
+			MigrationAdvanced { index: 1, blocks: 0 },
+			MigrationCompleted { index: 1, blocks: 1 },
+			MigrationAdvanced { index: 2, blocks: 0 },
+			MigrationAdvanced { index: 2, blocks: 1 },
+			MigrationCompleted { index: 2, blocks: 2 },
+			UpgradeCompleted,
+		]);
+	});
+}
+
+#[test]
+fn basic_works() {
+	test_closure(|| {
+		MigrationsStorage::set(vec![(SucceedAfter, 0), (SucceedAfter, 1), (SucceedAfter, 2)]);
+
+		System::set_block_number(1);
+		Migrations::on_runtime_upgrade();
+		run_to_block(10);
+
+		// Check that the executed migrations are recorded into `Historical`.
+		assert_eq!(
+			historic(),
+			vec![
+				mocked_id(SucceedAfter, 0),
+				mocked_id(SucceedAfter, 1),
+				mocked_id(SucceedAfter, 2),
+			]
+		);
+		// Check that we got all events.
+		assert_events(vec![
+			Event::UpgradeStarted { migrations: 3 },
+			Event::MigrationCompleted { index: 0, blocks: 1 },
+			Event::MigrationAdvanced { index: 1, blocks: 0 },
+			Event::MigrationCompleted { index: 1, blocks: 1 },
+			Event::MigrationAdvanced { index: 2, blocks: 0 },
+			Event::MigrationAdvanced { index: 2, blocks: 1 },
+			Event::MigrationCompleted { index: 2, blocks: 2 },
+			Event::UpgradeCompleted,
+		]);
+	});
+}
+
+#[test]
+fn failing_migration_keep_stuck_the_chain() {
```

Comment:
maybe this is better? 
```suggestion
fn failing_migration_sets_cursor_to_stuck() {
```